ships/0022-build-strategy-volumes.md

```diff
@@ -0,0 +1,179 @@
+
+### Goals
+
+- Build strategies can declare volumes whose data can be reused across steps.
+- Developers can use persistent volumes in a `BuildRun`.
+
+### Non-Goals
+
+- Securing volume usage/access with admission control.
+- Allow developers to inject arbitrary volumes into `BuildRun`s.
+
+## Proposal
+
+### User Stories
+
+#### Share information in build steps
+
+As a build strategy author
+I want to declare volumes in my build steps
+So that I can share information across build steps
+
+#### Use persistent volumes in build runs
+
+As a developer
+I want to declare persistent volume claims in my Builds
+So that I can cache resources like image layers and artifacts for subsequent BuildRuns
+
+### Implementation Notes
+
+The `BuildStrategy` API will be enhanced so that volumes can be declared alongside the build strategy steps.
+The volumes will provide fixed mount points that will be used in all steps, along with a default source for the underlying filesystem.
+Volumes can be marked `optional` indicating that the volume is not needed for a `BuildRun` to execute.
+Volumes can also indicate if the source can be overrode by a `Build` or `BuildRun`.
+
+Both the `Build` and `BuildRun` APIs will be enhanced to allow volumes to be specified.
+The volume name referenced in the `Build` or `BuildRun` must be declared in the build strategy.
+
+#### Sample API
+
+This is a sample of the API specified above.
+Field names and behavior are subject to change once this proposal is deemed implementable.
+
+`BuildStrategy`:
+
+```yaml
+metadata:
+  name: cached-docker-build
+spec:
+  buildSteps:
+  - name: build
+  ...
+  volumes:
+  - name: build-metadata
+    description: "Build metadata"
+    mountPath: /home/build/metadata
+    volumeSource:
+      emptyDir: {}
+  - name: image-cache
+    description: "Container image cache"
+    optional: true
+    mountPath: /var/lib/containers/cache
+    volumeSource:
+      emptyDir: {}
+    allowSourceOverride: true
+```
+
+`Build` and `BuildRun`:
+
+```yaml
+spec:
+  ...
+  strategy:
+    name: cached-docker-build
+    kind: BuildStrategy
+  volumes:
+  - name: image-cache
+    volumeSource:
+      persistentVolumeClaim:
+        name: pvc-image-cache
+```
+
+### Test Plan
+
+TBD
+
+### Release Criteria
+
+#### Removing a deprecated feature [if necessary]
+
+Not applicable
+
+#### Upgrade Strategy [if necessary]
+
+TBD
+
+### Risks and Mitigations
+
+Security is a concern with volumes, especially if arbitrary `HostPath` volume mounts are allowed in the API.
+This can be mitigated through admission controllers, which as a last resort can gate pod admission.
+Ideally we craft documentation which helps admins block admission at `Build` or `BuildRun` creation.
+
+## Drawbacks
+
+Volumes support adds complexity to the API that may require developers to fully inspect the referenced build strategies.
+We have assumed with other proposals that developers should not be expected to interact fully with `BuildStrategy` objects.
```

Comment:
Hmm ... interesting, I was thinking the opposite.  That the labels would be a way to associate a BuildVolume CRD with a particular build strategy.

i.e. you have 20 BuildVolumes.  2 are used by strategy A, 6 by strategy B.  Sure, maybe you could use naming conventions to draw associations.  But the value of a given label for a BuildVolume could be the strategy that uses it.

Separate point on the sharing across strategies.  I can see the notion of not sharing build volumes across strategies being the more likely landing spot.  But I'm not convinced that there will not be a non-trivial amount of image layer caching scenarios that span more than one strategy.  For example, I could see more than one maven based build strategy existing, and both want to leverage a volume that houses some very common maven downloaded jar files.

Reply:
Practically speaking, I don't think we want to design an API that encourages lots of build strategies to be created. Users shouldn't need to create separate build strategies for, say, kaniko because one uses an image layer cache, and another doesn't.

I'm going to add a more concrete example in the design.

Reply:
the brief write is fine with me @adambkaplan thanks